### DISCUSSION TITLE
Problems with Energy data from Tasmota on Sonoff POWR3

### DISCUSSION THREAD

**Discussion by wayner9** [arendst/Tasmota#20040] [General]
I have a Sonoff POWR3 connected to my hot tub heater.  When the heater is on it uses 5500W of power.  Tasmota shows that power usage.  But Tasmota is not properly calculating the energy used.  In the table below the first four columns come from parsing a Status 08 call to the POWR3 - see the JSON output of a "http://192.168.1.225/cm?cmnd=Status%208" command posted below.  The last two columns are my calculation of the change in 'Today' being reported by the device in each five minute increment, and a calculation of how much energy was actually used.

The correct power usage of 5500W is being reported. The readings are every five minutes.  Five minutes of 5500W of power consumes 0.46kWh which is 5500*(5/60)/1000.  But every five minutes the Energy Today is not incrementing nearly enough.  It is only incrementing by about 0.1kWh.  There is the question of calibration, but it appears that the calibration is for the power reading.  And the power reading is correct.  But what is incorrect is the Energy calculation which is power X time.

Anyone know what is going on?  Is it possible that there is a bug in the calculation of energy?  Maybe this bug only happens in higher powered devices?

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

</head>

<body link="#0563C1" vlink="#954F72">

Time | Status | Power | EnergyToday | Energy Change | Calc Energy
-- | -- | -- | -- | -- | --
2023-11-16T15:25:01 | ON | 0 | 0.55 | calc  | calc  
2023-11-16T15:30:00 | ON | 5498 | 0.642 | 0.092 | 0.458
2023-11-16T15:35:01 | ON | 5506 | 0.742 | 0.100 | 0.459
2023-11-16T15:40:00 | ON | 5492 | 0.84 | 0.098 | 0.458
2023-11-16T15:45:01 | ON | 5492 | 0.947 | 0.107 | 0.458
2023-11-16T15:50:00 | ON | 5521 | 1.052 | 0.105 | 0.460

</body>

</html>

```
{
    "StatusSNS": {
        "Time": "2023-11-16T16:18:08",
        "ENERGY": {
            "TotalStartTime": "2023-10-28T17:38:18",
            "Total": 8.918,
            "Yesterday": 1.171,
            "Today": 1.12,
            "Power": 0,
            "ApparentPower": 0,
            "ReactivePower": 0,
            "Factor": 0,
            "Voltage": 242,
            "Current": 0
        }
    }
}
```

**Comment by wayner9** [comment]
I just wanted to add - when I used this device with eWeLink prior to flashing with Tasmota it was reporting usage of about 5kWh/day which makes sense as the heater tends to run for about 30 minutes per day.

**Reply by sfromis** [reply]
Did you do the calibration, or did you just decide to ignore the suggestion of this?

It is more complex than just looking at "Power", as there are also power factor and apparent/reactive power. But the data you posted are not useful for seeing if something happens here.

**Reply by wayner9** [reply]
I didn't yet because the power is correct.  The issue is that I don't have a power meter that can measure that load.  I can assume that it is correct and type in 5500W for the power.  I have measured the voltage and it is 240V.  I have calibrated the voltage with the VoltageSet command.  I have turned up the heater and once it turns on I will apply the Power and Current settings.   

So you're saying that the Power measure can be correct but the energy measure can be incorrect?  That seems weird since energy is just power X time.  

But I will do this and see if it change anything.

**Reply by sfromis** [reply]
Calibration is all 3 commands/values, not just two. And probably do it in the stipulated order too. As suggested in the other thread, it may be workable to calculate ampere as watts / volts and multiply by 1000 for milliamps. **If** the tub heater is a *purely* resistive load (that's typically the case for heaters), calibration this way may not be perfect, but should get you reasonably close.

**Reply by wayner9** [reply]
I tried the commands although not initially in the exact order.  That didn't fix the problem, I am still only seeing energy usage of about 0.1kWh per five minute interval.  I did them again in the correct order but nothing has changed.  The Energy data is still wrong.  This should be pretty straightforward and I have to think that the energy reporting is derived from the power numbers.  And the power numbers are correct.  So the problem must be in the time interval used in the power calculation or in the formula.  Or maybe a bigger power number of 5500 for watts at 240V is causing issues.  And the energy was reported correctly under eWeLink, which probably doesn't mean anything.

For now I can do my own Energy calculation from the power and the time delta.

Where would I find the energy calcs in the Tasmota source code?

Here is the Tasmota console:
```
17:35:45.500 CMD: PowerSet 5500
17:35:45.507 MQT: stat/tasmota_509A6D/RESULT = {"PowerSetCal":5236}
17:36:03.594 CMD: VoltageSet 240
17:36:03.602 MQT: stat/tasmota_509A6D/RESULT = {"VoltageSetCal":1939}
17:36:25.463 CMD: CurrentSet 22.92
17:36:25.470 MQT: stat/tasmota_509A6D/RESULT = {"CurrentSetCal":16215}
```
And here is the power and energy data

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>





</head>

<body link="#0563C1" vlink="#954F72">


Time | Status | Power | EnergyToday | Energy Change | Calc Energy
-- | -- | -- | -- | -- | --
2023-11-16T17:15:00 | ON | 5403 | 1.18 |   |  
2023-11-16T17:20:01 | ON | 5403 | 1.273 | 0.093 | 0.450
2023-11-16T17:25:01 | ON | 5418 | 1.378 | 0.105 | 0.452
2023-11-16T17:30:01 | ON | 5389 | 1.475 | 0.097 | 0.449
2023-11-16T17:35:00 | ON | 5391 | 1.582 | 0.107 | 0.449
2023-11-16T17:40:00 | ON | 5500 | 1.69 | 0.108 | 0.458
2023-11-16T17:45:00 | ON | 5477 | 1.803 | 0.113 | 0.456
2023-11-16T17:50:01 | ON | 5483 | 1.921 | 0.118 | 0.457



</body>

</html>

**Reply by sfromis** [reply]
> CurrentSet 22.92

No the current is in milliampere, therefore I wrote to multiply by 1000.
